packages/cactus-api-client/src/main/typescript/socketio-api-client.ts

```diff
@@ -0,0 +1,345 @@
+        );
+
+        this.socket.on("connect_error", (err: Error) => {
+          this.log.error("##connect_error:", err);
+          this.socket.disconnect();
+          if (this.monitorSubject) {
+            this.monitorSubject.error(err);
+          }
+        });
+
+        this.socket.on("connect_timeout", (err: Record<string, unknown>) => {
+          this.log.error("####Error:", err);
+          this.socket.disconnect();
+          if (this.monitorSubject) {
+            this.monitorSubject.error(err);
+          }
+        });
+
+        this.socket.on("error", (err: Record<string, unknown>) => {
+          this.log.error("####Error:", err);
+          this.socket.disconnect();
+          if (this.monitorSubject) {
+            this.monitorSubject.error(err);
+          }
+        });
+
+        this.socket.on("monitor_error", (err: Record<string, unknown>) => {
+          this.log.error("#### Monitor Error:", err);
+          if (this.monitorSubject) {
+            this.monitorSubject.error(err);
+          }
+        });
+
+        this.socket.on("eventReceived", (res: any) => {
+          // output the data received from the client
+          this.log.debug("#[recv]eventReceived, res:", res);
+
+          this.checkValidator(this.options.validatorKeyPath, res.blockData)
+            .then((decodedData) => {
+              const resultObj = {
+                status: res.status,
+                blockData: decodedData.blockData,
+              };
+              this.log.debug("resultObj =", resultObj);
+              const event = new SocketLedgerEvent();
+              event.verifierId = this.options.validatorID;
+              this.log.debug(`##event.verifierId: ${event.verifierId}`);
+              event.data = resultObj;
+              if (this.monitorSubject) {
+                this.monitorSubject.next(event);
+              }
+            })
+            .catch((err) => {
+              this.log.error(err);
+            });
+        });
+
+        const emitStartMonitor = () => {
+          this.log.debug("##emit: startMonitor");
+          if (!monitorOptions || Object.keys(monitorOptions).length === 0) {
+            this.socket.emit("startMonitor");
+          } else {
+            this.socket.emit("startMonitor", monitorOptions);
+          }
+        };
+
+        if (this.socket.connected) {
+          emitStartMonitor();
+        } else {
+          this.socket.on("connect", () => {
+            this.log.debug("#connect");
+            emitStartMonitor();
+          });
+        }
+      } catch (err) {
+        this.log.error(`##Error: startMonitor, ${err}`);
+        this.monitorSubject.error(err);
+      }
+
+      return this.monitorSubject.pipe(
+        finalize(() => {
+          if (this.monitorSubject && !this.monitorSubject.observed) {
+            // Last observer finished
+            this.log.debug("##emit: stopMonitor");
+            this.socket.emit("stopMonitor");
```

Comment:
@outSH How do these string constants get disambiguiated on the back-end when there are multiple different endpoints that have different async use-cases?

Reply:
@petermetz There's only one single endpoint per "legacy" socketio validator, and all validators stop when they receive "stopMonitor", so there is no ambiguity (or I misunderstood your question)

Reply:
@outSH What I'm getting at is that I would assume that there are multiple endpoints if this is to be the `generic abstraction` as mentioned in the other comments of the originating issue and that causes ambiguity. You said in the other comment that this is a lift-and-shift which makes much more sense to me instead of describing it as a generic abstraction with a value add it's more like a lift-and-shift as-is. Which I am okay with as long as we call it what it is because this then feeds into later discussions about release management as well (which is why I'm having so many questions here - not to be annoying with small details).

Reply:
I don't recall calling `SocketIOApiClient` a generic abstraction, it is a component to communicate with socketio connectors with validator-style interface, as it is used in `cmd-socketio-server` at the moment, with as little change as possible. There's only one request<->response channel per connection, so there's no need to differentiate responses. Creating / deciding the common socketio interface has to be done in the near future (in other place), at the moment we want to support current validators as the way they worked so far. 

`ISocketApi` is the generic abstraction ensuring certain methods are available. How they are handled is managed by specific implementation (connection sharing and stuff), each ledger can implement different logic as long as they provide specified methods.

Reply:
> I don't recall calling SocketIOApiClient a generic abstraction,

@outSH IIRC it was in one of Takuma's comments describing the change in general.


> it is a component to communicate with socketio connectors with validator-style interface, as it is used in cmd-socketio-server at the moment, with as little change as possible. There's only one request<->response channel per connection, so there's no need to differentiate responses. Creating / deciding the common socketio interface has to be done in the near future (in other place), at the moment we want to support current validators as the way they worked so far.

> ISocketApi is the generic abstraction ensuring certain methods are available. How they are handled is managed by specific implementation (connection sharing and stuff), each ledger can implement different logic as long as they provide specified methods.

I understand. I feel like we are talking past each other though. I'll try from a different angle: Do you expect this (in it's current form) to be part of the 1.0 GA release early next year?

Reply:
> Do you expect this (in it's current form) to be part of the 1.0 GA release early next year?

The goal of my Japanese colleagues is to communicate `api-server` with current socketio-validator connectors, **without changing the current validator interface**, and also implement connector-agnostic `Verifier`. Both of these are done, one thing left is to upgrade socketio-validator connectors to socketio 4 so they can actually talk with `api-server` (which uses socketio 4), I plan to do this in separate PR by the end of this year. Beside that, all the goals are already met by this PR and from my point of view should be included in the GA.

Most of your comments were about interface between `SocketIoApiServer` and socketio-validator connector, which we can't change in near future, but I will create a separate github issue where I'd like to introduce few proposals, and start a discussion regarding the common socketio interface with all connectors. This also relates to naming `SocketIoApiServer`.

Reply:
@outSH Okay, cool. :-)